pkg/internal/apis/certmanager/v1alpha2/conversion.go

```diff
@@ -36,6 +36,16 @@ func Convert_v1alpha2_CertificateSpec_To_certmanager_CertificateSpec(in *v1alpha
 		out.Subject.Organizations = in.Organization
 	}
 
+	if in.KeyAlgorithm != "" || in.KeyEncoding != "" || in.KeySize != 0 {
+		if out.PrivateKey == nil {
+			out.PrivateKey = &certmanager.CertificatePrivateKey{}
+		}
+
+		out.PrivateKey.Algorithm = certmanager.PrivateKeyAlgorithm(in.KeyAlgorithm)
+		out.PrivateKey.Encoding = certmanager.PrivateKeyEncoding(in.KeyEncoding)
+		out.PrivateKey.Size = in.KeySize
+	}
+
```

Comment:
Does defaulting take place before or after this? If before, then those `in.` field ought never to be empty, right?
And if after, then if only one of them has a value, then all three values are set, which might prevent the defaults being applied?.

Reply:
We don't actually default right now at all, because the only way to properly default is via CRD/openapi annotations which we don't make use of at all currently.

But the path/order on *reads* if/when we do would be:

1) Read resource from etcd
2) Apply defaults for the API version *stored in etcd* (*not* the version that the user requested)
3) Convert into the 'request API version'

For writes, it will be:

1) User submits resource in 'request version'
2) Defaults are applied in 'request version'
3) Validation is performed in 'request version'
4) Conversion is performed to '*storage* version' (if required)
5) Object is persisted

At least, that's my understanding 😄 (cc @sttts or @liggitt?)

The reason we can't use mutating webhooks to apply defaults is that they are not executed on *reads* (i.e. step (2) of the read path above), which means if we add a new field with a default value, old resources that have already been persisted will not have this default value set when a client reads the object.